File: packages/agent-core/src/agent/permission/policies/browser-tool-ask.ts
import type {
  PermissionPolicy,
  PermissionPolicyContext,
  PermissionPolicyResult,
} from '../types';

export class BrowserToolAskPermissionPolicy implements PermissionPolicy {
  readonly name = 'browser-tool-ask';

  evaluate(context: PermissionPolicyContext): PermissionPolicyResult | undefined {
    if (!context.toolCall.name.startsWith('mcp__chrome-devtools__')) return;
    return {
      kind: 'ask',
      reason: { tool: context.toolCall.name },
    };
  }
}

File: rust-ody/crates/agent-rs/src/permission/policies/browser_tool_ask.rs
use crate::permission::types::{
    PermissionPolicy, PermissionPolicyContext, PermissionPolicyResolution,
};

pub struct BrowserToolAsk;

impl PermissionPolicy for BrowserToolAsk {
    fn name(&self) -> &str {
        "browser-tool-ask"
    }

    fn evaluate(
        &self,
        context: &PermissionPolicyContext<'_>,
    ) -> Option<PermissionPolicyResolution> {
        if context.tool_call.name.starts_with("mcp__chrome-devtools__") {
            let mut reason = std::collections::HashMap::new();
            reason.insert(
                "tool".to_string(),
                serde_json::json!(context.tool_call.name),
            );
            Some(PermissionPolicyResolution::Ask {
                reason: Some(reason),
                resolve_approval: None,
                resolve_error: None,
            })
        } else {
            None
        }
    }
}
